File: src/hubs/hub3_pharmacy_logistics/compounding_allergy_auditor.py
"""Compounding Allergy Auditor (Agent 10, Hub 3)."""

import hashlib
import os
import re
from datetime import datetime, timezone

SUBJECT_ID_PEPPER = os.environ.get("SUBJECT_ID_PEPPER", "CHANGE_ME_IN_PRODUCTION")

COMPOUND_ALLOWLIST = ["components", "allergy_profile"]

CROSS_REACTIVITY_MAP = {
    "penicillin": {"allergy_keywords": ["penicillin", "penicillins", "amoxicillin", "ampicillin"], "reaction_risk": "anaphylaxis"},
    "cephalosporin": {"allergy_keywords": ["cephalosporin", "cephalosporins", "cefazolin"], "reaction_risk": "anaphylaxis_cross_reactivity"},
    "sulfa": {"allergy_keywords": ["sulfa", "sulfonamide", "sulfonamides"], "reaction_risk": "anaphylaxis"},
}

_COMPONENT_NAME_PATTERN = re.compile(r"^([A-Za-z][A-Za-z-]*(?:\s+[A-Za-z][A-Za-z-]*)*)")


def hash_subject_id(mrn, dob):
    return hashlib.sha256(f"{SUBJECT_ID_PEPPER}:{mrn}:{dob}".encode()).hexdigest()


def mask_value(value):
    s = str(value) if value is not None else ""
    if len(s) <= 2:
        return "*" * len(s)
    return f"{s[0]}{'*' * (len(s) - 2)}{s[-1]}"


def iso_now():
    return datetime.now(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def normalize(s):
    s = re.sub(r"[^a-z]", "", str(s).lower())
    return re.sub(r"s$", "", s)


def extract_component_name(segment):
    trimmed = segment.strip()
    match = _COMPONENT_NAME_PATTERN.match(trimmed)
    return match.group(1).strip() if match else trimmed


class CompoundingAllergyAuditor:
    # Stage 1 (Formula Ingestion): split the compounding order into its
    # chemical component names and pair them with the patient's allergy
    # profile; the raw formula string is discarded after extraction.
    def ingest(self, formula_text, allergy_profile, patient_identifiers):
        if not isinstance(formula_text, str) or not formula_text.strip():
            raise TypeError("formula_text must be a non-empty string")
        if not isinstance(allergy_profile, list):
            raise TypeError("allergy_profile must be a list")
        patient_identifiers = patient_identifiers or {}
        mrn = patient_identifiers.get("mrn")
        dob = patient_identifiers.get("dob")
        if not mrn or not dob:
            raise ValueError("patient_identifiers requires mrn and dob to derive a subject_id")

        components = [c for c in (extract_component_name(seg) for seg in formula_text.split("+")) if c]

        record = {
            "subject_id": hash_subject_id(mrn, dob),
            "components": components,
            "allergy_profile": [str(a) for a in allergy_profile],
        }

        for key in list(record.keys()):
            if key != "subject_id" and key not in COMPOUND_ALLOWLIST:
                del record[key]
        return record

    # Stage 2 (Compliance & Cross-Reactivity Rules): defense-in-depth guard,
    # then normalize component and allergy names (handling plural mismatches
    # like "Penicillin" vs. "Penicillins") to find any cross-reactivity.
    def audit_cross_reactivity(self, record):
        leaked = next((k for k in record if k != "subject_id" and k not in COMPOUND_ALLOWLIST), None)
        if leaked:
            raise ValueError(f'Compliance violation: unexpected field "{leaked}" reached audit stage')

        conflicts = []
        for component in record["components"]:
            component_norm = normalize(component)
            map_entry = next(((key, info) for key, info in CROSS_REACTIVITY_MAP.items() if key in component_norm), None)
            if not map_entry:
                continue

            _, info = map_entry
            for allergy in record["allergy_profile"]:
                allergy_norm = normalize(allergy)
                is_match = any(normalize(keyword) == allergy_norm for keyword in info["allergy_keywords"])
                if is_match:
                    conflicts.append({"component": component, "allergy": allergy, "reaction_risk": info["reaction_risk"]})

        return {"conflicts": conflicts, "cleared": len(conflicts) == 0}

    # Stage 3 (Integration): only reachable when Stage 2 clears the order.
    # Shapes a cleanroom compound verification record.
    def build_cleanroom_verification(self, record):
        now = datetime.now(timezone.utc)
        return {
            "verification_type": "CleanroomCompoundVerification",
            "status": "verified",
            "subject_id": record["subject_id"],
            "components": record["components"],
            "allergy_profile_checked": record["allergy_profile"],
            "verification_id": f'CRV-{record["subject_id"][:10]}-{int(now.timestamp() * 1000)}',
            "verified_at": iso_now(),
        }

    # Stage 4a (Cleanroom Intercept): only reachable on a cross-reactivity
    # conflict — the verification payload is never generated on this path.
    def build_cleanroom_intercept(self, record, audit_result):
        return {
            "channel": "cleanroom_lab_tech_intercept",
            "priority": "hard_block",
            "subject_id": record["subject_id"],
            "headline": "STOP: anaphylactic cross-reactivity risk detected in compounding order",
            "conflicts": [
                f'Component "{c["component"]}" conflicts with documented allergy "{c["allergy"]}" — risk: {c["reaction_risk"]}.'
                for c in audit_result["conflicts"]
            ],
            "requested_action": "Do not compound as ordered. Contact the prescriber to select a non-cross-reactive alternative before proceeding.",
            "created_at": iso_now(),
        }

    # Stage 4b (Audit Telemetry): fully masked Splunk audit log — component
    # and allergy names are masked before they ever reach the log.
    def build_splunk_event(self, record, audit_result, outcome):
        masked_conflicts = [
            {"component": mask_value(c["component"]), "allergy": mask_value(c["allergy"]), "reaction_risk": c["reaction_risk"]}
            for c in audit_result["conflicts"]
        ]

        return {
            "time": datetime.now(timezone.utc).timestamp(),
            "host": "compounding-allergy-auditor",
            "source": "hub3_pharmacy_logistics/compounding_allergy_auditor",
            "sourcetype": "_json",
            "event": {
                "subject_id": record["subject_id"],
                "action": outcome,
                "cleared": audit_result["cleared"],
                "conflict_count": len(audit_result["conflicts"]),
                "masked_conflicts": masked_conflicts,
                "processed_at": iso_now(),
            },
        }

    def run(self, formula_text, allergy_profile, patient_identifiers):
        record = self.ingest(formula_text, allergy_profile, patient_identifiers)
        audit_result = self.audit_cross_reactivity(record)

        if audit_result["cleared"]:
            verification = self.build_cleanroom_verification(record)
            splunk_event = self.build_splunk_event(record, audit_result, "cleared_for_compounding")
            return {"verification": verification, "intercept": None, "splunk_event": splunk_event}

        intercept = self.build_cleanroom_intercept(record, audit_result)
        splunk_event = self.build_splunk_event(record, audit_result, "blocked_cross_reactivity")
        return {"verification": None, "intercept": intercept, "splunk_event": splunk_event}

File: src/hubs/hub3_pharmacy_logistics/compounding_allergy_auditor.js
'use strict';

const crypto = require('crypto');

const SUBJECT_ID_PEPPER = process.env.SUBJECT_ID_PEPPER || 'CHANGE_ME_IN_PRODUCTION';

const COMPOUND_ALLOWLIST = ['components', 'allergyProfile'];

const CROSS_REACTIVITY_MAP = {
  penicillin: {
    allergyKeywords: ['penicillin', 'penicillins', 'amoxicillin', 'ampicillin'],
    reactionRisk: 'anaphylaxis',
  },
  cephalosporin: {
    allergyKeywords: ['cephalosporin', 'cephalosporins', 'cefazolin'],
    reactionRisk: 'anaphylaxis_cross_reactivity',
  },
  sulfa: {
    allergyKeywords: ['sulfa', 'sulfonamide', 'sulfonamides'],
    reactionRisk: 'anaphylaxis',
  },
};

function hashSubjectId(mrn, dob) {
  return crypto
    .createHash('sha256')
    .update(`${SUBJECT_ID_PEPPER}:${mrn}:${dob}`)
    .digest('hex');
}

function maskValue(value) {
  const str = String(value ?? '');
  if (str.length <= 2) return '*'.repeat(str.length);
  return `${str[0]}${'*'.repeat(str.length - 2)}${str[str.length - 1]}`;
}

function normalize(str) {
  return String(str).toLowerCase().replace(/[^a-z]/g, '').replace(/s$/, '');
}

function extractComponentName(segment) {
  const trimmed = segment.trim();
  const match = trimmed.match(/^([A-Za-z][A-Za-z-]*(?:\s+[A-Za-z][A-Za-z-]*)*)/);
  return match ? match[1].trim() : trimmed;
}

class CompoundingAllergyAuditor {
  // Stage 1 (Formula Ingestion): split the compounding order into its
  // chemical component names and pair them with the patient's allergy
  // profile; the raw formula string is discarded after extraction.
  ingest(formulaText, allergyProfile, patientIdentifiers) {
    if (typeof formulaText !== 'string' || !formulaText.trim()) {
      throw new TypeError('formulaText must be a non-empty string');
    }
    if (!Array.isArray(allergyProfile)) {
      throw new TypeError('allergyProfile must be an array');
    }
    const { mrn, dob } = patientIdentifiers || {};
    if (!mrn || !dob) {
      throw new Error('patientIdentifiers requires mrn and dob to derive a subjectId');
    }

    const components = formulaText
      .split('+')
      .map(extractComponentName)
      .filter(Boolean);

    const record = {
      subjectId: hashSubjectId(mrn, dob),
      components,
      allergyProfile: allergyProfile.map(String),
    };

    for (const key of Object.keys(record)) {
      if (key !== 'subjectId' && !COMPOUND_ALLOWLIST.includes(key)) {
        delete record[key];
      }
    }
    return record;
  }

  // Stage 2 (Compliance & Cross-Reactivity Rules): defense-in-depth guard,
  // then normalize component and allergy names (handling plural mismatches
  // like "Penicillin" vs. "Penicillins") to find any cross-reactivity.
  auditCrossReactivity(record) {
    const leaked = Object.keys(record).find((key) => key !== 'subjectId' && !COMPOUND_ALLOWLIST.includes(key));
    if (leaked) {
      throw new Error(`Compliance violation: unexpected field "${leaked}" reached audit stage`);
    }

    const conflicts = [];
    for (const component of record.components) {
      const componentNorm = normalize(component);
      const mapEntry = Object.entries(CROSS_REACTIVITY_MAP).find(([key]) => componentNorm.includes(key));
      if (!mapEntry) continue;

      const [, info] = mapEntry;
      for (const allergy of record.allergyProfile) {
        const allergyNorm = normalize(allergy);
        const isMatch = info.allergyKeywords.some((keyword) => normalize(keyword) === allergyNorm);
        if (isMatch) {
          conflicts.push({ component, allergy, reactionRisk: info.reactionRisk });
        }
      }
    }

    return { conflicts, cleared: conflicts.length === 0 };
  }

  // Stage 3 (Integration): only reachable when Stage 2 clears the order.
  // Shapes a cleanroom compound verification record.
  buildCleanroomVerification(record) {
    return {
      verificationType: 'CleanroomCompoundVerification',
      status: 'verified',
      subjectId: record.subjectId,
      components: record.components,
      allergyProfileChecked: record.allergyProfile,
      verificationId: `CRV-${record.subjectId.slice(0, 10)}-${Date.now()}`,
      verifiedAt: new Date().toISOString(),
    };
  }

  // Stage 4a (Cleanroom Intercept): only reachable on a cross-reactivity
  // conflict — the verification payload is never generated on this path.
  buildCleanroomIntercept(record, auditResult) {
    return {
      channel: 'cleanroom_lab_tech_intercept',
      priority: 'hard_block',
      subjectId: record.subjectId,
      headline: 'STOP: anaphylactic cross-reactivity risk detected in compounding order',
      conflicts: auditResult.conflicts.map(
        (c) => `Component "${c.component}" conflicts with documented allergy "${c.allergy}" — risk: ${c.reactionRisk}.`
      ),
      requestedAction: 'Do not compound as ordered. Contact the prescriber to select a non-cross-reactive alternative before proceeding.',
      createdAt: new Date().toISOString(),
    };
  }

  // Stage 4b (Audit Telemetry): fully masked Splunk audit log — component
  // and allergy names are masked before they ever reach the log.
  buildSplunkEvent(record, auditResult, outcome) {
    return {
      time: Date.now() / 1000,
      host: 'compounding-allergy-auditor',
      source: 'hub3_pharmacy_logistics/compounding_allergy_auditor',
      sourcetype: '_json',
      event: {
        subjectId: record.subjectId,
        action: outcome,
        cleared: auditResult.cleared,
        conflictCount: auditResult.conflicts.length,
        maskedConflicts: auditResult.conflicts.map((c) => ({
          component: maskValue(c.component),
          allergy: maskValue(c.allergy),
          reactionRisk: c.reactionRisk,
        })),
        processedAt: new Date().toISOString(),
      },
    };
  }

  run(formulaText, allergyProfile, patientIdentifiers) {
    const record = this.ingest(formulaText, allergyProfile, patientIdentifiers);
    const auditResult = this.auditCrossReactivity(record);

    if (auditResult.cleared) {
      const verification = this.buildCleanroomVerification(record);
      const splunkEvent = this.buildSplunkEvent(record, auditResult, 'cleared_for_compounding');
      return { verification, intercept: null, splunkEvent };
    }

    const intercept = this.buildCleanroomIntercept(record, auditResult);
    const splunkEvent = this.buildSplunkEvent(record, auditResult, 'blocked_cross_reactivity');
    return { verification: null, intercept, splunkEvent };
  }
}

module.exports = {
  CompoundingAllergyAuditor,
  CROSS_REACTIVITY_MAP,
  COMPOUND_ALLOWLIST,
};
